src/printer_tree.c

```diff
@@ -0,0 +1,3107 @@
+    uint16_t lys_status;                    /**< Inherited status CURR, DEPRC, OBSLT. */
+    uint16_t lys_config;                    /**< Inherited config W or R. */
+    const struct lysp_node_list *last_list; /**< The last LYS_LIST passed. */
+};
+
+/**
+ * @brief Main structure for browsing the libyang tree
+ */
+struct trt_tree_ctx {
+    trt_actual_section section;         /**< To which section pn points. */
+    const struct lys_module *module;    /**< Schema tree structures. */
+    const struct lysp_node *pn;         /**< Actual pointer to parsed node. */
+    const struct lysp_node *tpn;        /**< Pointer to actual top-node. */
+};
+
+/**
+ * @brief Used for updating trt_tree_ctx
+ */
+struct trt_tree_ctx_node_patch {
+    const struct lysp_node *pn;         /**< Actual pointer to parsed node. */
+    const struct lysp_node *tpn;        /**< Pointer to actual top-node. */
+};
+
+/** Getter function for tro_lysp_node_charptr function. */
+typedef const char *(*trt_get_charptr_func)(const struct lysp_node *pn);
+
+/******************************************************************************
+ * Definitions start
+ *****************************************************************************/
+/******************************************************************************
+ * Definition of the general Trg functions
+ *****************************************************************************/
+
+/**
+ * @brief Print a substring but limited to the maximum length.
+ * @param[in] str is pointer to source.
+ * @param[in] len is number of characters to be printed.
+ * @param[in,out] out is output handler.
+ * @return str parameter shifted by len.
+ */
+static const char *
+trg_print_substr(const char *str, size_t len, struct ly_out *out)
+{
+    for (size_t i = 0; i < len; i++) {
+        ly_print_(out, "%c", str[0]);
+        str++;
+    }
+    return str;
+}
+
+/**
+ * @brief Pointer is not NULL and does not point to an empty string.
+ * @param[in] str is pointer to string to be checked.
+ * @return 1 if str pointing to non empty string otherwise 0.
+ */
+static ly_bool
+trg_charptr_has_data(const char *str)
+{
+    return str != NULL && str[0] != '\0';
+}
+
+/**
+ * @brief Check if 'word' in 'src' is present where words are delimited by 'delim'.
+ * @param[in] src is source where words are separated by delim.
+ * @param[in] word to be searched.
+ * @param[in] delim is delimiter between words in src.
+ * @return 1 if src contains word otherwise 0.
+ */
+static ly_bool
+trg_word_is_present(const char *src, const char *word, char delim)
+{
+    if ((src == NULL) || (src[0] == '\0') || (word == NULL)) {
+        return 0;
+    }
+
+    const char *hit = strstr(src, word);
+
+    if (hit) {
+        /* word was founded at the begin of src
+         * OR it match somewhere after delim
+         */
+        if ((hit == src) || (*(hit - 1) == delim)) {
```

Comment:
Please use `[]` when accessing `char *` array (`hit[-1]`). You can use this for all other cases but we generally prefer to use `[]` for all array types.